### ISSUE TITLE
firebase_auth onAuthStateChanged not updated on linkWithCredential

### ISSUE THREAD

**Issue by erf** [FirebaseExtended/flutterfire#2690]
**Describe the bug**
I'm linking an Anonymous firebaseUser to a Phone provider using `firebaseUser.linkWithCredential` but `onAuthStateChanged` is not triggered, even if i call `firebaseUser.reload()`.

**To Reproduce**
1. Sign in with an anonymous user
2. Link to a phone provider using `linkWithCredential`

**Expected behavior**
I would expect `Firebase.instance.onAuthStateChanged` to be triggered with a new event, especially after calling `firebaseUser.reload`.

**Additional context**

**Flutter doctor**
Run `flutter doctor` and paste the output below:

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.17.1, on Mac OS X 10.15.4 19E287, locale en-GB)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
[✓] Android Studio (version 4.0)
[✓] VS Code (version 1.40.0)
[✓] Connected device (1 available)

• No issues found!
```

**Comment by TahaTesser** [comment]
Hi @erf 
Does this look similar to https://github.com/FirebaseExtended/flutterfire/issues/2344?

**Comment by erf** [comment]
@TahaTesser I am able to link an anonymous user to a phone provider, but the `FirebaseAuth.instance.onAuthStateChanged` was not called. 

..but after some more testing, i think the problem is related to `CombineLatestStream` of `rxdart`, in that it does not push a new event when `onAuthStateChanged` does.

I'll close this for now, but if you have any experience with this scenario, please help.

Thank you.

**Comment by erf** [comment]
@TahaTesser Actually it does seem somewhat random, like he mentions in #2344. I removed the use of `CombineLatestStream`, and then first auth was called with "phone" provider removed after i `unlinkFromProvider` for all providers in `providerData` and then call `reload`, but later when i did it, the "phone" provider was still there.

**Comment by erf** [comment]
I think this is related to https://github.com/FirebaseExtended/flutterfire/issues/717

**Comment by erf** [comment]
For anyone else with this problem:

I think i will solve this by adding a `rx.SubjectBehaviour authSubject`, i will then listen to `onAuthStateChanged` and add events to `authSubject`.

I will also add a separate method:
```
  void forceAuthStreamUpdate() async {
    authSubject.add(await FirebaseAuth.instance.currentUser());
  }
```

For cases where `onAuthStateChanged` is not called.. until this issue is fixed by the Flutter Team.

**Comment by TahaTesser** [comment]
Hi @erf 
Can you please provide Your `flutter run --verbose` and a minimal complete reproducible code sample.
Thank you

**Comment by erf** [comment]
@TahaTesser that could make quite the complex example

Could you tell me if the following:
```
firebaseUser.linkWithCredential
firebaseUser.reload().
```
is supposed to trigger an event in `FirebaseAuth.instance.onAuthStateChanged` with a  `FirebaseUser` with updated `providerData` ?

If not, then i guess everything is fine. Except i think this is what should happen.

**Comment by RWOverdijk** [comment]
I know this repo is not for the firebase JS web sdk, but I have the same issue there. I'm signed in as an anonymous user and upon linking with a credential it does not trigger the onAuthStateChanged.

Normally I wouldn't comment on an unrelated repo but I think these issues might be related somehow. .reload() also doesn't do anything for me so that also seems the same.

**Comment by Ehesp** [comment]
Reload doesn't retrigger anything, you should be able to call `reload` and access the `currentUser` after it.

**Comment by RWOverdijk** [comment]
I get that, I was talking about triggering the auth state change.

**Comment by bulwinkel** [comment]
For other who end up here, based on comments in this issue https://github.com/FirebaseExtended/flutterfire/issues/717 there will be `userChanges` stream available in a future release of firebase_auth as part of the work being done in https://github.com/FirebaseExtended/flutterfire/issues/2582 but it is unclear in which release it will become available.

**Comment by schankam** [comment]
I'm having the same problem after linking an anonymous user with an other provider, the onAuthStateChanged is not triggered. I am using a StreamProvider with FirebaseUser, and was expecting it to be triggered so my UI is refreshed accordingly (currently using a Visibility Widget to display extra option with the firebaseUser.isAnonymous property). If this is not working correctly, how are we supposed to rely on that boolean ?

**Comment by erf** [comment]
I believe this is solved in firebase_auth 0.18.0. You can now listen to the `userChanges` stream for these events.
